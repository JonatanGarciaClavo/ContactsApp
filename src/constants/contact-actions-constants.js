export const LOADING_CONTACT = 'LOADING_CONTACT';
export const INITIALIZE_CONTACT = 'INITIALIZE_CONTACT';
export const UPDATE_CONTACT_ATTRIBUTE = 'UPDATE_CONTACT_ATTRIBUTE';
export const VALIDATE_CONTACT_ATTRIBUTE = 'VALIDATE_CONTACT_ATTRIBUTE';
export const VALIDATE_CONTACT = 'VALIDATE_CONTACT';
export const REQUEST_CONTACT = 'REQUEST_CONTACT';
export const REQUEST_SAVE_CONTACT = 'REQUEST_SAVE_CONTACT';
