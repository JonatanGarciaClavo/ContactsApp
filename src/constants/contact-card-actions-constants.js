export const LOADING_CONTACT_CARD = 'LOADING_CONTACT_CARD';
export const INITIALIZE_CONTACT_CARD = 'INITIALIZE_CONTACT_CARD';
export const REQUEST_CONTACT_CARD = 'REQUEST_CONTACT_CARD';
export const REQUEST_DELETE_CONTACT_CARD = 'REQUEST_DELETE_CONTACT_CARD';
export const INITILIZE_CONTACT_CARD_FROM_OTHER_VIEW = 'INITILIZE_CONTACT_CARD_FROM_OTHER_VIEW';
