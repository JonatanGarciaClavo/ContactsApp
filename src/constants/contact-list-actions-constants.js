export const LOADING_CONTACT_LIST = 'LOADING_CONTACT_LIST';
export const INITIALIZE_CONTACT_LIST = 'INITIALIZE_CONTACT_LIST';
export const CHANGE_LIST_MODE = 'CHANGE_LIST_MODE';
