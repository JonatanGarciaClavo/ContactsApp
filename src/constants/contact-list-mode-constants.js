export const LIST_MODE = 'list';
export const CARD_MODE = 'card'
