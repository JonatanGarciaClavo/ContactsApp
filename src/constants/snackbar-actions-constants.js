export const CLOSE_ERROR_MESSAGE = 'CLOSE_ERROR_MESSAGE';
export const SET_ERROR_MESSAGE = 'SET_ERROR_MESSAGE';
