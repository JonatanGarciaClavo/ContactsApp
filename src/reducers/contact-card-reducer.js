import Immutable from 'immutable';
import { REQUEST_CONTACT_CARD, REQUEST_DELETE_CONTACT_CARD, REQUEST_CONTACT_CARD_SUCCESS,
  RESET_CONTACT_CARD } from '../constants/contact-card-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';

export const initialState = new Immutable.Map({
  loading: false,
  contact: new Immutable.Map(),
});

const contactCard = (state = initialState, action) => {
  switch (action.type) {
    case (REQUEST_CONTACT_CARD):
    case (REQUEST_DELETE_CONTACT_CARD):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_CARD_SUCCESS):
      return state
        .set('contact', action.contact)
        .set('loading', false);
    case (RESET_CONTACT_CARD):
      return initialState;
    default:
      return state;
  }
};

export default contactCard;
