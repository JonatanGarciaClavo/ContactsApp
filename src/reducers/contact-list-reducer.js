import { Record, List } from 'immutable';
import { REQUEST_CONTACT_LIST, REQUEST_CONTACT_LIST_SUCCESS, CHANGE_LIST_MODE, RESET_CONTACT_LIST }
  from '../constants/contact-list-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';
import { LIST_MODE } from '../constants/contact-list-mode-constants';

export const ContactList = new Record({
  loading: false,
  contacts: new List(),
  orderBy: 'name',
  mode: LIST_MODE,
});

const contactList = (state = new ContactList(), action) => {
  switch (action.type) {
    case (REQUEST_CONTACT_LIST):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_LIST_SUCCESS):
      return state
        .set('contacts', action.contacts)
        .set('loading', false);
    case (CHANGE_LIST_MODE):
      if (action.mode === state.get('mode')) {
        return state;
      }
      return state.set('mode', action.mode || LIST_MODE);
    case (RESET_CONTACT_LIST):
      return new ContactList();
    default:
      return state;
  }
};

export default contactList;
