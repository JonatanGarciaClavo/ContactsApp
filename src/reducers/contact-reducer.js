import _ from 'lodash';
import { Record, Map, fromJS } from 'immutable';
import { REQUEST_CONTACT, REQUEST_SAVE_CONTACT, REQUEST_CONTACT_SUCCESS, UPDATE_CONTACT_ATTRIBUTE,
  VALIDATE_CONTACT, VALIDATE_CONTACT_ATTRIBUTE, RESET_CONTACT,
  } from '../constants/contact-actions-constants';
import { SET_ERROR_MESSAGE } from '../constants/snackbar-actions-constants';
import ContactModel from '../models/ContactModel';

const requiredValidation = value =>
  value !== undefined && value !== null && (!_.isEmpty(value) || _.isNumber(value));

const isEmail = (value) => {
  if (value) {
    const regExp = /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i;// eslint-disable-line
    return regExp.test(value);
  }
  return true;
};

const isUrl = (value) => {
  if (value) {
    const regExp = new RegExp("^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!10(?:\.\d{1,3}){3})(?!127(?:\.\d{1,3}){3})(?!169\.254(?:\.\d{1,3}){2})(?!192\.168(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]+-?)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:\/[^\s]*)?");// eslint-disable-line
    return regExp.test(value);
  }
  return true;
};

const applyValidator = (ruleObject, value) => {
  let error;
  let functionValidator;
  if (_.isFunction(ruleObject.rule)) {
    functionValidator = ruleObject.rule;
  }
  if (functionValidator && !functionValidator(value)) {
    error = ruleObject.error || `${ruleObject.rule} not followed`;
  } else if (!functionValidator) {
    console.warn('applyValidator: %s does not exists', ruleObject.rule);// eslint-disable-line
  }
  return error;
};

const validateAll = (rulesObject, contact) => {
  const errors = {};

  _.each(rulesObject, (rules, name) => {
    if (rules && _.isArray(rules)) {
      const value = contact[name];
      _.each(rules, (ruleObject) => {
        const error = applyValidator(ruleObject, value);
        if (error) {
          errors[name] = error;
        }
      });
    } else {
      console.warn('validateAll: missing parameters rules ');// eslint-disable-line
    }
  });

  return _.isEmpty(errors) ? false : errors;
};

const rulesObject = {
  name: [{ rule: requiredValidation, error: 'Name is required' }],
  email: [
    { rule: requiredValidation, error: 'Email is required' },
    { rule: isEmail, error: 'Email format is incorrect' },
  ],
  imgUrl: [{ rule: isUrl, error: 'Invalid url' }],
};

export const Contact = new Record({
  loading: false,
  isModified: false,
  contact: new ContactModel(),
  errors: new Map(),
});


const contact = (state = new Contact(), action) => {
  const { name, value } = action;
  let error;
  let errors;
  switch (action.type) {
    case (REQUEST_SAVE_CONTACT):
    case (REQUEST_CONTACT):
      return state.set('loading', true);
    case (SET_ERROR_MESSAGE):
      return state.set('loading', false);
    case (REQUEST_CONTACT_SUCCESS):
      return state
        .set('contact', action.contact || new ContactModel())
        .set('loading', false)
        .set('errors', new Map());
    case (UPDATE_CONTACT_ATTRIBUTE):
      return state
        .set('isModified', true)
        .setIn(['contact', name], value);
    case VALIDATE_CONTACT_ATTRIBUTE:
      error = validateAll(_.pick(rulesObject, name), { [name]: value });
      if (error && error[name]) {
        return state.setIn(['errors', name], error[name]);
      }
      return state.deleteIn(['errors', name]);
    case VALIDATE_CONTACT:
      errors = validateAll(rulesObject, state.get('contact').toJS());
      return state.set('errors', new Map(fromJS(errors || {})));
    case RESET_CONTACT:
      return new Contact();
    default:
      return state;
  }
};

export default contact;
