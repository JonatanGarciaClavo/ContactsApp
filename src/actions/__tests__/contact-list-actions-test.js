import mockStore from 'redux-mock-store';
import { recieveContactList, requestContactList, loadingContactList, deleteContact, editContact,
  loadContactCard, changeListMode } from '../contact-list-actions';
import { initialState } from '../../reducers/contact-list-reducer';
import { CARD_MODE, LIST_MODE } from '../../constants/contact-list-mode-constants';

const store = mockStore({ contactList: initialState });

beforeEach(() => {
  store.clearActions();
});

it('should handle requestContactList action', () => {
  store.dispatch(requestContactList());
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle loadingContactList action', () => {
  store.dispatch(loadingContactList());
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle recieveContactList action', () => {
  store.dispatch(recieveContactList([]));
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle deleteContact action', () => {
  store.dispatch(deleteContact('-1'));
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle editContact action', () => {
  store.dispatch(editContact({
    id: '-1',
  }));
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle loadContactCard action', () => {
  store.dispatch(loadContactCard({
    id: '-1',
  }));
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle changeListMode action LIST_MODE', () => {
  store.dispatch(changeListMode(LIST_MODE));
  expect(store.getActions()).toMatchSnapshot();
});
it('should handle changeListMode action CARD_MODE', () => {
  store.dispatch(changeListMode(CARD_MODE));
  expect(store.getActions()).toMatchSnapshot();
});
