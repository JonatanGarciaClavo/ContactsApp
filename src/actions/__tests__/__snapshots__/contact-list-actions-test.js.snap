exports[`test should handle changeListMode action CARD_MODE 1`] = `
Array [
  Object {
    "mode": "card",
    "type": "CHANGE_LIST_MODE",
  },
]
`;

exports[`test should handle changeListMode action LIST_MODE 1`] = `Array []`;

exports[`test should handle deleteContact action 1`] = `
Array [
  Object {
    "id": "-1",
    "type": "DELETE_CONTACT",
  },
]
`;

exports[`test should handle editContact action 1`] = `
Array [
  Object {
    "contact": Object {
      "id": "-1",
    },
    "type": "INITILIZE_CONTACT_FROM_OTHER_VIEW",
  },
]
`;

exports[`test should handle loadContactCard action 1`] = `
Array [
  Object {
    "contact": Object {
      "id": "-1",
    },
    "type": "INITILIZE_CONTACT_CARD_FROM_OTHER_VIEW",
  },
]
`;

exports[`test should handle loadingContactList action 1`] = `
Array [
  Object {
    "type": "LOADING_CONTACT_LIST",
  },
]
`;

exports[`test should handle recieveContactList action 1`] = `
Array [
  Object {
    "contacts": Array [],
    "type": "INITIALIZE_CONTACT_LIST",
  },
]
`;

exports[`test should handle requestContactList action 1`] = `
Array [
  Object {
    "type": "REQUEST_CONTACT_LIST",
  },
]
`;
